airflow/jobs.py

```diff
@@ -1644,26 +1738,31 @@ def _execute_helper(self, processor_manager):
                 # If a task instance is up for retry but the corresponding DAG run
                 # isn't running, mark the task instance as FAILED so we don't try
                 # to re-run it.
-                self._change_state_for_tis_without_dagrun(simple_dag_bag,
-                                                          [State.UP_FOR_RETRY],
-                                                          State.FAILED)
+                ret = self._change_state_for_tis_without_dagrun(simple_dag_bag,
+                                                                [State.UP_FOR_RETRY],
+                                                                State.FAILED)
+                if not ret:
+                    continue
                 # If a task instance is scheduled or queued, but the corresponding
                 # DAG run isn't running, set the state to NONE so we don't try to
                 # re-run it.
-                self._change_state_for_tis_without_dagrun(simple_dag_bag,
-                                                          [State.QUEUED,
-                                                           State.SCHEDULED],
-                                                          State.NONE)
-
+                ret = self._change_state_for_tis_without_dagrun(simple_dag_bag,
+                                                                [State.QUEUED,
+                                                                 State.SCHEDULED],
+                                                                State.NONE)
+                if not ret:
+                    continue
                 self._execute_task_instances(simple_dag_bag,
                                              (State.SCHEDULED,))
 
             # Call heartbeats
             self.log.info("Heartbeating the executor")
             self.executor.heartbeat()
 
-            # Process events from the executor
-            self._process_executor_events(simple_dag_bag)
+            # Process all events from the executor
+            full_simple_dag_bag = SimpleDagBag(self.full_simple_dags.values())
+            self.executor_events = self._process_executor_events(full_simple_dag_bag)
```

Comment:
This has already been introduced in Airflow 2.0. https://github.com/apache/airflow/pull/7597

Reply:
> Your change is targeted at Airflow 1.9, is that correct? Plus, please describe your change.

**Background**: Our team has been using Airflow 1.9 for two years. We use it as the scheduler of our deploy system, but when the number of dags we need to schedule at the same time exceeds 300, the task scheduling delay of Dag is no longer stable, sometimes More than 1 minute.
So I made improvements and fixes to the 1.9 version of Airflow. The detailed improvement information is as follows (same as the [update log](https://github.com/apache/airflow/pull/8259/files/edc04002ea6605b529088dc4d58e8631bec47577#diff-bb7ee25332d5fdca3ead2f545d72b359))：
1. Added a feedback event mechanism to the scheduler to make the task scheduling delay stable at 3s;
2. Solved the problem of db deadlock caused by worker and scheduler updating task_instance at the same time (this will cause the scheduler to exit abnormally);
3. In order to improve the availability of the scheduler, I used distributed locks to add a hot standby function to the scheduler.

Reply:
> This has already been introduced in Airflow 2.0. #7597

You mean the ruturn value of `self._process_executor_events` ?
In my change,  the return value of `self._process_executor_events` contain execute result  of upsteam tasks, and I parse the result, then set Dag of tasks to head of schedule queue, improve the scheduling speed of downstream tasks of Dag .

Reply:
> > Your change is targeted at Airflow 1.9, is that correct? Plus, please describe your change.
> 
> **Background**: Our team has been using Airflow 1.9 for two years. We use it as the scheduler of our deploy system, but when the number of dags we need to schedule at the same time exceeds 300, the task scheduling delay of Dag is no longer stable, sometimes More than 1 minute.
> So I made improvements and fixes to the 1.9 version of Airflow. The detailed improvement information is as follows (same as the [update log](https://github.com/apache/airflow/pull/8259/files/edc04002ea6605b529088dc4d58e8631bec47577#diff-bb7ee25332d5fdca3ead2f545d72b359))：
> 
> 1. Added a feedback event mechanism to the scheduler to make the task scheduling delay stable at 3s;
> 2. Solved the problem of db deadlock caused by worker and scheduler updating task_instance at the same time (this will cause the scheduler to exit abnormally);
> 3. In order to improve the availability of the scheduler, I used distributed locks to add a hot standby function to the scheduler.

Hi, @buhuipao I know what you said, but pls take a look at Jarek said above. 
> We want to encourage people to update to the latest released 1.10 version rather than fix the old version.
User would use latest release instead of 1.9, so I think we should target to master branch. And I check your UPDATING.md, maybe we should separate PR which easy to review, WDYT

Reply:
BTW, I don't know Baidu use Airflow before, do you mind to add your company and your team members in https://github.com/apache/airflow/blob/master/README.md

Reply:
Thank for your advice，I will push right code to the master branch in next time.